nextcord/http.py

```diff
@@ -124,25 +140,329 @@ def bucket(self) -> str:
+            this number.
+        """
+        _log.debug("Bucket %s: Resetting after %s seconds.", self.bucket, time)
+        await asyncio.sleep(time)
+        self.remaining = self.limit
+        self._on_reset_event.set()
+        _log.debug("Bucket %s: Reset, allowing requests to continue.", self.bucket)
 
-    def __enter__(self) -> Self:
-        return self
+    @property
+    def migrating(self) -> Optional[str]:
+        """If not ``None``, this indicates what bucket acquiring requests should migrate to."""
+        return self._migrating
+
+    def migrate_to(self, bucket: str) -> None:
+        """Signals to acquiring requests, both present and future, that they need to migrate to a new bucket."""
+        self._migrating = bucket
+        self.remaining = self.limit
+        self._on_reset_event.set()
+        _log.debug(
+            "Bucket %s: Deprecating, acquiring requests will migrate to a new bucket.", bucket
+        )
 
-    def defer(self) -> None:
-        self._unlock = False
+    async def __aenter__(self) -> None:
+        await self.acquire()
 
-    def __exit__(
+    async def __aexit__(
         self,
-        exc_type: Optional[Type[BE]],
-        exc: Optional[BE],
-        traceback: Optional[TracebackType],
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
     ) -> None:
-        if self._unlock:
-            self.lock.release()
+        self.release()
+
+    @property
+    def locked(self) -> bool:
+        return self.remaining <= 0
+
+    async def acquire(self) -> bool:
+        # If no more requests can be made but the event is set, clear it.
+        if self.locked and self._on_reset_event.is_set():
+            _log.debug(
+                "Bucket %s: Hit the remaining request limit of %s, locking until reset.",
+                self.bucket,
+                self.limit,
+            )
+            self._on_reset_event.clear()
+            if not self.resetting:
+                self.start_reset_task()
+
+        # Waits in a loop for the event to be set, clearing the event as needed and looping.
+        while not self._on_reset_event.is_set():
+            _log.debug("Bucket %s: Not set yet, waiting for it to be set.", self.bucket)
+            await self._on_reset_event.wait()
+
+            if self.locked and self._on_reset_event.is_set():
+                _log.debug(
+                    "Bucket %s: Hit the remaining limit of %s, locking until reset.",
+                    self.bucket,
+                    self.limit,
+                )
+                self._on_reset_event.clear()
+                if not self.resetting:
+                    self.start_reset_task()
+
+        if self.migrating:
+            raise RateLimitMigrating(
+                f"This RateLimit is deprecated, you need to migrate to bucket {self.migrating}"
+            )
```

Comment:
Now I know its easy to get via RateLimit.migrating, but would make sense to have the new bucket hash on the error?